Scan from purl

san-zrl opened the pull request:
PR for running a scan from a Package URL:
- Purl can be specified in giturl field in UI.
- Purl must contain a `@version` tag
- Branch if specified is overridden by `@version` tag from purl
- Subfolder is ignored
- Purl resolved via deps.dev. Process fails if deps.dev doesn't know about it.
- Prepared fetching the CBOM from github (result of envisioned github action). Command is a noop for the moment.
- CBOMReadModel: Purl is identifier for CBOM, resolved giturl stored as repository.

n1ckl0sk0rtge commented on src/main/java/com/ibm/domain/scanning/ResolvedScanRequest.java:
```diff
@@ -0,0 +1,70 @@
+        return gitUrl;
+    }
+
+    public void setGitUrl(ScanUrl gitUrl) {
+        this.gitUrl = gitUrl;
+    }
```
I think, the `ResolvedScanRequest` could be a `record` instead of a class, since value objects are per definition immutable ([see](https://martinfowler.com/bliki/ValueObject.html)).

san-zrl replied:
`ResolvedScanRequest` removed.

n1ckl0sk0rtge commented on src/main/java/com/ibm/domain/scanning/ResolvedScanRequest.java:
```diff
@@ -0,0 +1,70 @@
+import jakarta.annotation.Nullable;
+
+public class ResolvedScanRequest implements IValueObject {
+    @Nullable ScanUrl gitUrl;
```
```suggestion
    @Nullable ScanUrl gitUrl;
```
I would remove the giturl here

n1ckl0sk0rtge commented on src/main/java/com/ibm/domain/scanning/ScanAggregate.java:
```diff
@@ -35,45 +40,70 @@
+    // private static final Logger LOGGER = LoggerFactory.getLogger(ScanAggregate.class);
 
-    @Nonnull private final ScanRequest scanRequest;
+    @Nonnull private ResolvedScanRequest request;
```
```suggestion
    @Nonnull private final ScanRequest scanRequest;
```
would keep the request as is

san-zrl replied:
Kept request as is and added new mutable state attributes (`@Nullable purl`, `@Nullable gitUrl`, `@Nonnull revision`) to `ScanAggregate`.

n1ckl0sk0rtge commented on src/main/java/com/ibm/domain/scanning/ScanAggregate.java:
```diff
@@ -35,45 +40,70 @@
 
-    @Nonnull private final ScanRequest scanRequest;
+    @Nonnull private ResolvedScanRequest request;
     @Nullable private Commit commit;
```
```suggestion
    @Nullable private GitUrl giturl;
    @Nullable private Purl purl;
    @Nullable private Commit commit;
```

san-zrl replied:
Kept request as is and added new mutable state attributes (@Nullable purl, @Nullable gitUrl, @Nonnull revision) to ScanAggregate.

n1ckl0sk0rtge commented on src/main/java/com/ibm/domain/scanning/ScanAggregate.java:
```diff
@@ -35,45 +40,70 @@
     private ScanAggregate(@Nonnull final ScanId id, @Nonnull final ScanRequest scanRequest) {
         super(id, new ArrayList<>());
-        this.scanRequest = scanRequest;
+        this.request = new ResolvedScanRequest(scanRequest);
+        if (request.scanUrl.isPurl()) {
+            try {
+                PackageURL purl = new PackageURL(request.scanUrl.value());
+                if (purl.getVersion() != null) {
+                    this.request.setRevision(new Revision(purl.getVersion()));
+                }
+            } catch (MalformedPackageURLException e) {
+                // ignore: scanRequest ha been validated
+            }
+        } else {
+            request.setGitUrl(request.scanUrl);
+        }
```
```suggestion
        this.scanRequest = scanRequest;
        if (request.scanUrl.isPurl()) {
            try {
                PackageURL purl = new PackageURL(request.scanUrl.value());
                if (purl.getVersion() != null) {
                    this.purl  = purl;
                    // this object should already allow you to do things like
                    // purl.getVersion()
                    // purl.getName()
                    // ...
                }
            } catch (MalformedPackageURLException e) {
                // ignore: scanRequest ha been validated
            }
        } else {
            this. gitUrl = request.scanUrl;
        }
```

san-zrl replied:
```java
        this.scanRequest = scanRequest;
        try {
            this.purl = new PackageURL(scanRequest.scanUrl().value());
            this.revision = new Revision(this.purl.getVersion());
        } catch (MalformedPackageURLException e) {
            this.gitUrl = new GitUrl(scanRequest.scanUrl().value());
            this.revision = scanRequest.revision();
        }
```

n1ckl0sk0rtge commented on src/main/java/com/ibm/domain/scanning/ScanAggregate.java:
```diff
@@ -35,45 +40,70 @@
     private ScanAggregate(
             @Nonnull ScanId id,
             @Nonnull ScanRequest scanRequest,
+            @Nullable ScanUrl gitUrl,
```
I think this has to be changes when applying the changes from above

san-zrl replied:
I reactivated the old GitUrl class to represent the gitUrl in ScanAggregate which is passed in here.

n1ckl0sk0rtge commented on src/main/java/com/ibm/infrastructure/database/readmodels/CBOMReadRepository.java:
```diff
@@ -75,7 +76,7 @@ public CBOMReadRepository(@Nonnull IDomainEventBus domainEventBus) {
 
     @Override
-    public @Nonnull Optional<CBOMReadModel> findBy(@Nonnull GitUrl gitUrl) {
+    public @Nonnull Optional<CBOMReadModel> findBy(@Nonnull ScanUrl gitUrl) {
```
```suggestion
    public @Nonnull Optional<CBOMReadModel> findBy(@Nonnull Purl purl) {
```
should be "findeble" by gitUrl or PURL

n1ckl0sk0rtge commented on src/main/java/com/ibm/infrastructure/scanning/repositories/Scan.java:
```diff
@@ -60,6 +61,8 @@ class Scan extends PanacheEntityBase {
     @Nullable public String commitHash;
     @Nullable public String subFolder;
+    @Nullable public String purl;
+    @Nullable public String tag;
```
Do we still need to explicitly store the tag here?

san-zrl replied:
We don't. I store branch/tag in `revision`.

n1ckl0sk0rtge commented on src/main/java/com/ibm/usecases/compliance/queries/RequestComplianceCheckForScannedGitRepositoryQueryHandler.java:
```diff
@@ -69,7 +69,7 @@ public RequestComplianceCheckForScannedGitRepositoryQueryHandler(
                 compliancePreparationService.receiveCryptographicAssets(
                         this.readRepository,
-                        new GitUrl(requestComplianceCheckForScannedGitRepositoryQuery.gitUrl()),
+                        new ScanUrl(requestComplianceCheckForScannedGitRepositoryQuery.gitUrl()),
```
For compliance either a `giturl` or `purl` should be the required input. A `scanUrl` should not be existing in the "compliance" domain

san-zrl replied:
I reactivated the old `GitUrl` class to represent the gitUrl in `ScanAggregate` which is passed in here.

n1ckl0sk0rtge commented on src/main/java/com/ibm/usecases/compliance/service/CompliancePreparationService.java:
```diff
@@ -42,9 +42,9 @@ public final class CompliancePreparationService {
     public Collection<CryptographicAsset> receiveCryptographicAssets(
             @Nonnull ICBOMReadRepository readRepository,
-            @Nonnull GitUrl gitUrl,
+            @Nonnull ScanUrl gitUrl,
```
Same as "For compliance either a giturl or purl should be the required input. A scanUrl should not be existing in the "compliance" domain"

san-zrl replied:
I reactivated the old GitUrl class to represent the gitUrl in ScanAggregate which is passed in here.

n1ckl0sk0rtge reviewed (changes requested):
Some changes required. See comments

san-zrl commented:
I removed `ResolvedScanRequest` completely and added new mutable state attributes (`@Nullable purl`, `@Nullable gitUrl`, `@Nonnull revision`) to `ScanAggregate`. The original ScanRequest and also its ScanUrl remains unchanged. `giturl` is set either directly (if the ScanUrl was a git link) or via purl resolution (if the ScanUrl was a purl). In the latter case `purl` is populated. Likewise, `revision` is either set to the git branch (if the ScanUrl was a git link) or to the purl version. This separates purl and gitUrl cleanly, and any of them is never used instead of the other.

In `CbomReadModel` `repository` is always the giturl, `projectIdentifier is the scan url (the purl or the giturl identifier). There are now findBy methods for both `GitUrl` and 'PackageURL` parameters (with and without commit parameter). Not sure how many of them we need.

I also changed the IO to reflect our enhancements. The projectIdentifier is now used in the list of scans and in the scan result (instead of the gitUrl). On the scan result the full gitUrl is shown as a label.
